rotary/rules/polish.smk

```diff
@@ -15,37 +15,120 @@ rule prepare_medaka_polish_input:
+        """
+
+checkpoint generate_contig_manifest:
+    input:
+        "{sample}/{step}/medaka_input/{sample}_input.fasta"
+    output:
+        contig_manifest="{sample}/{step}/medaka/{sample}_contigs.txt"
+    conda:
+        "../envs/medaka.yaml"
+    log:
+        "{sample}/logs/{step}/contig_names_medaka.log"
     benchmark:
         "{sample}/benchmarks/{step}/medaka.txt"
+    shell:
+        """
+        seqkit seq -n {input} > {output.contig_manifest} 2>> {log}
+        """
+
+rule polish_contig_medaka:
+    input:
+        calls_to_draft_bam='{sample}/{step}/medaka/calls_to_draft.bam',
+        calls_to_draft_bam_index='{sample}/{step}/medaka/calls_to_draft.bam.bai',
+        contig_manifest="{sample}/{step}/medaka/{sample}_contigs.txt"
+    output:
+        contig_polished=temp('{sample}/{step}/medaka/results/{sample}_{contig}.hd5')
+    conda:
+        "../envs/medaka.yaml"
+    log:
+        "{sample}/logs/{step}/medaka_{sample}_{contig}.log"
+    benchmark:
+        "{sample}/benchmarks/{step}/medaka_{sample}_{contig}.txt"
     params:
         medaka_model=config.get("medaka_model"),
-        batch_size=config.get("medaka_batch_size")
+        batch_size=config.get("medaka_batch_size"),
+        medaka_results_dir='{sample}/{step}/medaka/results',
+    threads:
+        2
+    shell:
+        """
+        medaka consensus {input.calls_to_draft_bam} \
+          {params.medaka_results_dir}/{wildcards.sample}_{wildcards.contig}.hd5 \
+          --model {params.medaka_model} \
+          --batch {params.batch_size} \
+          --threads {threads} \
+          --region {wildcards.contig} > {log} 2>&1
+        """
+
+def aggregate_medaka_polished_contigs(wildcards):
+    """
+    Callback function that generates a list contig HDF5 files that will be needed for rule stitch_medaka.
+
+    :param wildcards: These are the wildcards present in rule stich_medaka.
+    :return: HDF5 files to be generated for each contig by multiple executions of rule polish_contig_medaka.
+    """
+    # Force execution of checkpoint generate_contig_files.
+    # Passes wildcards from rule stitch_medaka to generate_contig_files.
+    # Execution will generate a contig manifest file and revaluate the DAG.
+    contig_manifest_path = checkpoints.generate_contig_manifest.get(**wildcards).output[0]
+
+    # Open the resulting contig manifest file and read in the contig names.
+    with contig_manifest_path.open() as contig_file:
+        contigs_names = [line.strip() for line in contig_file]
+
+    # Returns the expected paths to the per-contig HDF5 polishing files.
+    return expand("{sample}/{step}/medaka/results/{sample}_{contig}.hd5",sample=wildcards.sample,
+        step=wildcards.step,contig=contigs_names)
+
+rule stitch_medaka:
+    input:
+        hdf5s=aggregate_medaka_polished_contigs,
+        draft_fasta_dir="{sample}/polish/medaka_input/",
```

Comment:
Just checking: is this directory input variable redundant now? It seems like it is not called in the shell section, and I couldn't think of another reason why it might need to be there.

Reply:
So the long-read mapping generates `.mmi` and `.fai` files in the same folder as the draft FASTA (e.g., `{sample}/polish/medaka_input/{sample}_input.fasta.mmi`).

```python
rule prepare_medaka_polish_input:
    input:
        "{sample}/assembly/{sample}_assembly.fasta"
    output:
        fasta=temp("{sample}/polish/medaka_input/{sample}_input.fasta"),
        fasta_dir=temp(directory("{sample}/polish/medaka_input/"))
```

In `prepare_medaka_polish_input,` the draft FASTA was initially set as a temp(). So, the FASTA would be deleted but not the `mmi` and `fai` files generated in that folder by `map_long_reads_for_medaka.` One can't directly reference the `fai` and `mmi` files in `prepare_medaka_polish_input` because they haven't been created yet. You can't delete them in the `map_long_reads_for_medaka` because they don't have a `step` wild card. So I found a way to delete the entire `"{sample}/polish/medaka_input` directory by setting it to temp in `prepare_medaka_polish_input.` If you don't reference the directory later in the DAG, the entire folder will be deleted immediately after the rule runs, deleting the draft FASTA. So, I referenced the draft FASTA directory as input into `stitch_medaka`; this prevents the directory from being deleted until the polishing is complete.

Reply:
Does that work or do you think there is a better way to do it?

Reply:
Ah, I see! Thanks for the detailed explanation. Yes, the way you've written the code makes sense to keep the `medaka_input` dir as temp.

It might help to add brief a comment above rule `stitch_medaka` to clarify why the `medaka_input` dir is declared as an input, so that someone doesn't delete this input line during code cleanup in the future. Otherwise, I am OK to keep the code as-is!

Reply:
Agreed